File: go/example_agent.go
// A2A Go SDK - Example Agent
//
// This example demonstrates how to create an A2A-enabled agent in Go.
//
// Run:
//
//	go run example_agent.go
package main

import (
	"fmt"
	"log"

	"github.com/mentessaas/a2a-protocol/go/a2a"
)

// SimpleTaskHandler handles tasks for the agent
func SimpleTaskHandler(action string, input map[string]interface{}, sender string) map[string]interface{} {
	fmt.Printf("📥 Received task: action=%s from=%s\n", action, sender)

	switch action {
	case "echo":
		return map[string]interface{}{
			"echo": input["message"],
		}
	case "add":
		a := input["a"].(float64)
		b := input["b"].(float64)
		return map[string]interface{}{
			"result": a + b,
		}
	default:
		return map[string]interface{}{
			"error": "Unknown action",
		}
	}
}

func main() {
	// Create an agent
	agent := a2a.NewAgent(
		"calculator-agent",
		"Calculator Agent",
		[]string{"math", "calculate", "add", "subtract"},
	)

	// Register with directory (optional)
	err := agent.Register("http://localhost:9001", "http://localhost:8080")
	if err != nil {
		log.Printf("Registration: %v", err)
	}

	// Start the server
	fmt.Println("🚀 Starting A2A agent server...")
	err = a2a.RunServer(
		"calculator-agent",
		"Calculator Agent",
		[]string{"math", "calculate", "add", "subtract"},
		9001,
		SimpleTaskHandler,
	)
	if err != nil {
		log.Fatal(err)
	}
}

// Example of using the agent client (uncomment to use)
/*
func clientExample() {
	agent := a2a.NewAgent("my-agent", "My Agent", []string{"search"})

	// Discover agents
	other, _ := agent.Discover([]string{"calculator"}, "http://localhost:8080")
	if other != nil {
		// Send a task
		result, err := agent.SendTask(other.AgentID, "add", map[string]interface{}{
			"a": 10,
			"b": 20,
		}, "http://localhost:8080")
		if err != nil {
			log.Fatal(err)
		}

		output, _ := json.Marshal(result.Output)
		fmt.Printf("Result: %s\n", output)
	}
}
*/

File: rust/examples/example_agent.rs
//! A2A Protocol Rust Example Agent
//!
//! Run with: cargo run --example example_agent

use a2a::{A2AAgent, run_server};
use serde_json::json;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Create and register an agent
    let mut agent = A2AAgent::new(
        "calculator-agent",
        "Calculator Agent",
        vec!["math".to_string(), "calculate".to_string(), "add".to_string()],
    );

    // Register with directory (optional)
    if let Err(e) = agent.register("http://localhost:9001", "http://localhost:8080").await {
        println!("Registration note: {}", e);
    }

    // Start server with task handler
    run_server(
        "calculator-agent",
        "Calculator Agent",
        vec!["math".to_string(), "calculate".to_string(), "add".to_string()],
        9001,
        |action, input, sender| {
            println!("📥 Received task: action={} from={}", action, sender);
            
            match action.as_str() {
                "add" => {
                    let a = input["a"].as_f64().unwrap_or(0.0);
                    let b = input["b"].as_f64().unwrap_or(0.0);
                    json!({"result": a + b})
                }
                "echo" => {
                    json!({"echo": input["message"]})
                }
                _ => {
                    json!({"error": "Unknown action"})
                }
            }
        },
    ).await?;

    Ok(())
}

// Example of client usage
/*
async fn client_example() -> Result<(), Box<dyn std::error::Error>> {
    let agent = A2AAgent::new(
        "my-agent",
        "My Agent",
        vec!["search".to_string()],
    );

    // Discover agents
    let other = agent.discover(
        vec!["calculator".to_string()],
        "http://localhost:8080"
    ).await?;

    if let Some(other) = other {
        // Send a task
        let result = agent.send_task(
            &other.agent_id,
            "add",
            json!({"a": 10, "b": 20}),
            "http://localhost:8080"
        ).await?;

        println!("Result: {:?}", result);
    }

    Ok(())
}
*/
